[Fix #1785] Handle timeout for network failure in utils.utils/http-get and http-post

foopang opened the pull request:
Fix #1785 using [react-native-fetch-polyfill](https://github.com/robinpowered/react-native-fetch-polyfill)

status: ready <!-- Can be ready or wip -->

foopang committed:
[Fix #1785] Handle timeout for network failure in utils.utils/http-get and http-post

Signed-off-by: Julien Eluard <julien.eluard@gmail.com>

jeluard commented:
@foopang Do you mind adding a small test validating the timeout behavior?

@foopang @yenda Also I am wondering if we could integrate the polyfill in our codebase, given how small and finished it is. It would make one dependency less and simplify things for other plateforms (e.g. `status-desktop`)

rasom commented on src/status_im/utils/utils.cljs:
```diff
@@ -62,8 +62,9 @@
-                                       :headers {"Cache-Control" "no-cache"}}))
+   (-> (fetch url (clj->js {:method  "GET"
+                            :headers {"Cache-Control" "no-cache"}
+                            :timeout 3000}))
```
Let's extract `3000` to a constant

foopang commented:
@jeluard Added a small test to validate the timeout behavior. For the requirement of integrating the polyfill in our codebase, should I just put the file in `resources/js/vendors`?

yenda commented:
@jeluard are you sure about vendoring this polyfil ? It will probably evolve with fetch version in the future

jeluard commented:
@yenda No, thinking out loud. We can probably revisit that later. @foopang Let's keep it as a dependency for now.

yenda commented on test/appium/tests/test_wallet.py:
```diff
@@ -120,3 +120,18 @@ def test_eth_and_currency_balance(self):
         wallet.verify_currency_balance(eth_rate, errors)
         assert not errors, 'errors occurred:\n{}'.format('\n'.join(errors))
+
+    @pytest.mark.wallet
```
@antdanchenko any opinion on this appium test ?

jeluard replied:
Ping @antdanchenko

antdanchenko replied:
thank you for adding the test, much appreciated! 
however, since we are running automated tests on cloud infrastructure, there could be some issues with turning network connection off, but lets give it a try

jeluard replied:
@antdanchenko Could you get more feedback on that point? Or can you only get some after it's merged?

lukaszfryc replied:
I think we should not have such tests at the end-end level. Instead, a unit test triggering a failure and checking what method is called (`show-popup "Error"`) or something like that will probably be enough. 

As @antdanchenko said, it probably will not work in the cloud (saucelabs).

lukaszfryc replied:
I've also run it locally and it didn't work correctly. No error is shown because device checks tokens prices after going offline (`self.driver.set_network_connection(1)`) which means the request is not being sent at all.

lukaszfryc replied:
@foopang let me rework this test for you. I will extend it a bit and make sure it works in our cloud. In the meantime you can focus on the issues I mentioned in my previous comment. Let me know how it goes. Thanks!

oskarth commented:
@jeluard @antdanchenko @yenda Please finish review with either yes or no with reservations.

jeluard commented:
@foopang Is that a connection timeout or request timeout? Also what happens when the timeout happens?

jeluard reviewed (changes requested):
Provide documentation about the change and its behavior.

foopang commented:
@jeluard I think it is a connection timeout as described:

> React Native's XMLHttpRequest interface exposes a timeout property sent to the RCTNetworking module, as well as an abort method. fetch does not expose access to this by default, this polyfill allows specifying a timeout within the options.
>
> This value attached to NSMutableURLRequest, where the native networking layer will enforce the timeout rule.
>
> The result of the timeout being reached will result in a promise rejected with a `TypeError('Network rqeuest failed').

BTW do you want `http-get` and `http-post` to accept `timeout-ms` as a function argument so that it doesn't have to be a hard-coded value?

yenda commented:
@foopang yes taking a timeout -ms function argument would be nice.

yenda commented on package.json:
```diff
@@ -58,6 +58,7 @@
     "react-native-drawer-layout": "2.0.0",
     "react-native-emoji-picker": "git+https://github.com/status-im/react-native-emoji-picker.git",
     "react-native-fcm": "10.0.3",
+    "react-native-fetch-polyfill": "^1.1.2",
```
please don't use the ^

foopang commented:
@yenda updated, please have a look.

yenda commented on package.json:
```diff
@@ -57,6 +57,7 @@
     "react-native-dialogs": "0.0.20",
     "react-native-emoji-picker": "git+https://github.com/status-im/react-native-emoji-picker.git",
     "react-native-fcm": "10.0.3",
+    "react-native-fetch-polyfill": "^1.1.2",
```
remove the ^ please

foopang replied:
fixed.

oskarth commented:
@jeluard @yenda Please re-review and if you still disapprove, write what changes you are expecting @foopang to do for PR to be accepted.

@foopang Please resolve conflicts (sorry a lot of them, not sure why exactly).

jeluard commented on src/status_im/ui/screens/events.cljs:
```diff
@@ -106,21 +106,20 @@
+(defn- http-get [{:keys [url response-validator success-event-creator failure-event-creator timeout-ms]}]
+  (let [on-success #(re-frame/dispatch (success-event-creator %))
+        on-error   #(re-frame/dispatch (success-event-creator %))]
+    (cond
```
We might benefit from passing args to `utils/http-get` as a map here.

foopang replied:
@jeluard I only set the last argument a map which will take `:valid-response?` and `:timeout-ms` as optional data, as I assume `url` `on-success` `on-error` are all required, so not putting inside the map. Does it make sense to you?

jeluard commented on src/status_im/utils/utils.cljs:
```diff
@@ -2,7 +2,10 @@
             [status-im.i18n :as i18n]
             [clojure.string :as str]
-            [status-im.react-native.js-dependencies :as rn-dependencies]))
+            [status-im.react-native.js-dependencies :as rn-dependencies :refer [fetch]]))
```
As a general rule we use `as`, not `refer`.

jeluard commented:
@foopang Looks generally good, requested some stylistic changes. Thanks and sorry for taking a long time to get back to you!

foopang commented:
@jeluard Updated. Please have a look.

jeluard commented on src/status_im/ui/screens/events.cljs:
```diff
@@ -108,21 +108,20 @@
+  (let [on-success #(re-frame/dispatch (success-event-creator %))
+        on-error   #(re-frame/dispatch (failure-event-creator %))
+        opts       (cond-> {}
+                     response-validator (assoc :valid-response? response-validator)
```
Could be simplified with a static map? Then nil values could be handled in `utils/http-get`

jeluard commented:
@foopang Looks good!

janherich commented:
@foopang Please resolve conflicts

foopang commented:
@jeluard @janherich Updated and resolved conflicts.

jeluard commented on src/status_im/ui/screens/events.cljs:
```diff
@@ -109,21 +109,19 @@
-                    #(re-frame/dispatch (failure-event-creator %)))))
+  (fn [{:keys [action data success-event-creator failure-event-creator timeout-ms]}]
+    (let [on-success #(re-frame/dispatch (success-event-creator %))
+          on-error   #(re-frame/dispatch (failure-event-creator %))]
```
Could we do something similar to `http-get` and have last argument a map? This way we can also use a static map (and no cond).
Sorry missed that one!

foopang replied:
@jeluard fixed now.

jeluard commented:
@foopang Thanks!
@yenda Does that look great to you?

goranjovic commented:
@rasom Are you happy with the changes? Can we move it forward?

lukaszfryc commented:
@foopang I've found couple of issues while testing on Samsung Galaxy S8, Android 7.0:

1. The error does not appear in Transaction History for timeout in `GET https://ropsten.etherscan.io/api?module=account&action=txlist&address={address}`. What's more, if there is a timeout the screen freezes showing a spinner and there is simply no way to get out of this state. None of the buttons work:

<img src="https://user-images.githubusercontent.com/7532782/36596775-1fd12b30-18a7-11e8-944f-75ea6db9cad2.png" width="300">

Note that the same scenario works nice for Wallet.

2. The error does not look good. I'm wondering if there is a better way to communicate it. Maybe we could use the "Offline" bar we show in Home but make it red? In the current solution "!" is not centered and there seems to be too much space above it. @yenda what do you think?

<img src="https://user-images.githubusercontent.com/7532782/36596270-600b520e-18a5-11e8-888b-5d4c7e2736ef.png" width="300">

3. I think we need to come up with a better message. "Error loading data" maybe too technical for our users. How about something like "Oops, there was an issue with loading your data". @yenda any thoughts?

lukaszfryc reviewed (dismissed):
@foopang please check my comments

status-open-bounty commented:
![Contract deploying](https://openbounty.status.im/img/deploying_contract.png)
Transaction: [0x75a7c5e6514cd60ab5f7d914774f2934a04c1588b8e41a6cab92b4e02a678ac5](https://etherscan.io/tx/0x75a7c5e6514cd60ab5f7d914774f2934a04c1588b8e41a6cab92b4e02a678ac5)

foopang commented:
@lukaszfryc Thanks for your review! I fixed these problems you found:
- The error does not appear in Transaction History for timeout
- if there is a timeout the screen freezes showing a spinner
- "!" is not centered and there seems to be too much space above it

![screenshot from 2018-02-25 17-47-00](https://user-images.githubusercontent.com/4666757/36640324-1f8c3be0-1a57-11e8-9973-a716a6a52c28.png)
![screenshot from 2018-02-25 17-46-49](https://user-images.githubusercontent.com/4666757/36640326-2390e81c-1a57-11e8-8827-48c43241011f.png)

I also removed the appium test as you mentioned that it will not probably work in the cloud. But I don't think I can add a unit test for that as it seems `http-get` / `http-post` won't work without a real device / simulator.

@jeluard @yenda  Please note that I have to undo some changes you made in previous commits in order to get the loading error showing.

jeluard commented:
@foopang Great work! 

Sorry I missed that before but we don't display error specifically anymore in wallet. While waiting forr a generic error handling UI we just display basic `alert` popup. 
Could you rollback this part to what it was before your rebase? Thanks!

jeluard commented:
@foopang Could you resolve the conflict?

foopang commented:
@jeluard Updated and added `alert` popup. Let me know if this is fine for you now.

jeluard commented:
@foopang Great work thanks!
